internal/batch/batch_processor.go

```diff
@@ -101,103 +130,249 @@ func (bw *batchWork) estimateSize() int64 {
 	return sizeEstimate
 }
 
-// The assemblyLoop accepts work into the pipe as quickly as possible.
-// It dispatches work asynchronously to the persistenceLoop, which is responsible for
-// calling back each piece of work once persisted into a batch
-// (doesn't wait until that batch is sealed/dispatched).
-// The assemblyLoop seals batches when they are full, or timeout.
+func (bp *batchProcessor) status() *ProcessorStatus {
+	bp.statusMux.Lock()
+	defer bp.statusMux.Unlock()
+	return &ProcessorStatus{
+		Dispatcher: bp.conf.dispatcherName,
+		Name:       bp.conf.name,
+		Status:     bp.flushStatus, // copy
+	}
+}
+
+func (bp *batchProcessor) newAssembly(initalWork ...*batchWork) {
+	bp.assemblyID = fftypes.NewUUID()
+	bp.assemblyQueue = append([]*batchWork{}, initalWork...)
+	bp.assemblyQueueBytes = batchSizeEstimateBase
+}
+
+// addWork adds the work to the assemblyQueue, and calculates if we have overflowed with this work.
+// We check for duplicates, and add the work in sequence order.
+// This helps in the case for parallel REST APIs all committing to the DB at a similar time.
+// With a sufficient batch size and batch timeout, the batch will still dispatch the messages
+// in DB sequence order (although this is not guaranteed).
+func (bp *batchProcessor) addWork(newWork *batchWork) (full, overflow bool) {
+	newQueue := make([]*batchWork, 0, len(bp.assemblyQueue)+1)
+	added := false
+	skip := false
+	// Check it's not in the recently flushed lish
+	for _, flushedSequence := range bp.flushedSequences {
+		if newWork.msg.Sequence == flushedSequence {
+			log.L(bp.ctx).Debugf("Ignoring add of recently flushed message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+			skip = true
+			break
+		}
+	}
+	// Build the new sorted work list, checking there for duplicates too
+	for _, work := range bp.assemblyQueue {
+		if newWork.msg.Sequence == work.msg.Sequence {
+			log.L(bp.ctx).Debugf("Ignoring duplicate add of message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+			skip = true
+		}
+		if !added && !skip && newWork.msg.Sequence < work.msg.Sequence {
+			newQueue = append(newQueue, newWork)
+			added = true
+		}
+		newQueue = append(newQueue, work)
+	}
+	if !added && !skip {
+		newQueue = append(newQueue, newWork)
+		added = true
+	}
+	if added {
```

Comment:
`added` will _always_ be true here unless `skip` was set. Again just feels like a lot of the logic should be skipped early if `skip` gets flipped to true.

Reply:
this is a good point 👍

Reply:
That's not actually true, if (as is the most common case) we're adding to the end of the list.

Reply:
I will do a simplification experiment, to see if I can make the add logic based on if the item is > than the current one... which I think would always add

Reply:
Yea, in the common case `added = true` will be hit up above on line 180, so it will be true by this point.

Reply:
... that went badly - just leaving the skip part

Reply:
ok - early return was the solution 👍  (`overflow` and `full` can be assumed to be `false` if we didn't add anything)